Python/bytecodes.c

```diff
@@ -2024,13 +2024,11 @@ dummy_func(
             // Combined: COMPARE_OP (float ? float) + POP_JUMP_IF_(true/false)
             DEOPT_IF(!PyFloat_CheckExact(left), COMPARE_OP);
             DEOPT_IF(!PyFloat_CheckExact(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
             double dleft = PyFloat_AS_DOUBLE(left);
             double dright = PyFloat_AS_DOUBLE(right);
-            // 1 if <, 2 if ==, 4 if >; this matches when _to_jump_mask
-            int sign_ish = 2*(dleft > dright) + 2 - (dleft < dright);
-            DEOPT_IF(isnan(dleft), COMPARE_OP);
-            DEOPT_IF(isnan(dright), COMPARE_OP);
-            STAT_INC(COMPARE_OP, hit);
+            // 1 if NaN, 2 if <, 4 if >, 8 if ==; this matches when_to_jump_mask
+            int sign_ish = 1 << (2 * (dleft >= dright) + (dleft <= dright));
```

Comment:
Credit to @markshannon for figuring this out. I had tried something similar with `<=` and `>=`, but couldn't get the masks to work correctly without his help.

Reply:
This is potentially a bit dangerous: a comparison with NaN may raise FE_INVALID. It may be safer to use the `is_less` macro instead.

From C99 (§7.12.14.3):

> The isless macro determines whether its first argument is less than its second argument. The value of isless(x, y) is always equal to (x) < (y); however, unlike (x) < (y), isless(x, y) does not raise the ‘‘invalid’’ floating-point exception when x and y are unordered.

Reply:
Sorry, that should be `islessequal` of course, not `isless`.

Reply:
`float_richcompare` uses simple `i < j` comparisons. Why is that OK, but not here?

Reply:
Hm, interesting. I had no idea that floating-point comparisons like these could crash. Is it a signaling NaN thing, or something more subtle and implementation-specific?

I agree with @markshannon that we should do whatever `float` does for now, and change them both later if it’s deemed unsafe.

(Also, the current code already has the same safety issue, if so. So it’s not like this PR is making the situation any worse.)

Reply:
@brandtbucher No, not specific to signalling NaNs, just to comparisons involving NaNs. And yes, `float_richcompare` should be updated to use the safer code, now that we're allowed to depend on C99 features.

Reply:
Okay. I'll merge this, and create a separate issue where we can discuss how/if we should avoid potential floating-point exceptions everywhere (since that seems like a much bigger change).

Reply:
And yes, it's platform-specific. I think we're in a place now where all current platforms do the "right" thing, but that wasn't always the case. There were a few related issues in the past - #81655 is the only one I'm finding right now, though. So yes, probably a non-issue at this point.

Reply:
> create a separate issue

Thanks. Don't worry about this, though - I think I'm fussing about things that were historically an issue, but are no longer. It's not that long ago that we _did_ protect all floating-point operations, e.g., here's `float_richcompare` in Python 2.7: https://github.com/python/cpython/blob/ca079a3ea30098aff3197c559a0e32d42dda6d84/Objects/floatobject.c#L600-L621

But it seems that all current platforms do follow the IEEE 754 standard's advice on "default" floating-point exception handling settings, so this is something we no longer need to worry about.

Reply:
(FTR, it looks like Python 3.9 is where we discarded the `PyFPE_START_PROTECT` / `PyFBE_END_PROTECT` guards everywhere. They're still present in 3.8.)